eng/targets/ResolveReferences.targets

```diff
@@ -91,24 +88,55 @@
         @(Reference->WithMetadataValue('IsSharedSource', 'true'));
         @(Reference->WithMetadataValue('PrivateAssets', 'All'))" />
     <_OriginalReferences Include="@(Reference)" />
+  </ItemGroup>
 
-    <!--
-      Turn Reference items into a ProjectReference when UseProjectReferences is true.
-      Order matters. This comes before package resolution because projects should be used when possible instead of packages.
+  <!--
+      Turn Reference items into a ProjectReference when UseProjectReferences is true. Order matters; this
+      comes before package resolution because projects should be used when possible instead of packages.
     -->
-    <_ProjectReferenceByAssemblyName Condition="'$(UseProjectReferences)' == 'true'"
-      Include="@(ProjectReferenceProvider)"
-      Exclude="@(_UnusedProjectReferenceProvider)" />
-
-    <ProjectReference Include="@(_ProjectReferenceByAssemblyName->'%(ProjectPath)')" />
-    <Reference Remove="@(_ProjectReferenceByAssemblyName)" />
+  <ItemGroup Condition=" '$(EnableCustomReferenceResolution)' == 'true' AND '$(UseProjectReferences)' == 'true' ">
+    <!-- Copy then Update / Copy to intersect the ProjectReferenceProvider and Reference item groups. -->
+    <_AllProjectReference Include="@(ProjectReferenceProvider)" />
+
+    <!-- Use only Reference items when project is a reference provider. Simplifies project moves and shortens files. -->
+    <_AllProjectReference Update="@(ProjectReference->'%(Filename)')" DirectUse="1" />
+
+    <_AllProjectReference Update="@(Reference)" Use="1">
```

Comment:
Is there a guarantee we'll never need to update this list of metadata?

Reply:
There are few guarantees in life😈 But, the current list goes pretty far beyond what we're likely to use in `*.csproj` and `*.fsproj` projects and should hold us in good stead until and unless new metadata is invented. I included everything I saw used in existing `@(Reference)` and `@(ProjectReference)` items in this repo as well as what seemed plausible from the documentation.